lib/Dialect/Moore/MooreTypes.cpp

```diff
@@ -197,143 +127,31 @@ namespace moore {
 namespace detail {
 struct IntTypeStorage : TypeStorage {
   using KeyTy = unsigned;
-  using Kind = IntType::Kind;
 
-  IntTypeStorage(KeyTy key) : kind(static_cast<Kind>(key)) {}
-  static KeyTy pack(Kind kind) { return static_cast<unsigned>(kind); }
-  bool operator==(const KeyTy &key) const { return pack(kind) == key; }
+  IntTypeStorage(KeyTy key)
+      : width(key >> 1), domain(static_cast<Domain>((key >> 0) & 1)) {}
```

Comment:
May I ask what the role played by the `>>` and `<<`?   Just my personal question, I searched on Groq and ChatGPT, but I still don't understand it. Thanks! :smile:

Reply:
The width and domain are packed into a single unsigned int: "domain" is in the lowest bit, and "width" in the upper 31 bits (the `pack` function does that). This unpacks the unsigned int again: width is set to bits 31:1 of the key, and domain is set to bit 0 of the key.

Reply:
I understand it. Because of `pack` function. Thanks!!! 😃